### PR TITLE
service logs

### PR DESCRIPTION
**Pull request by aluzzardi**
This PR adds `docker service logs` support by wiring up the SwarmKit logs plumbing.

This includes:
- Container Adapter & Controller: To push up the logs from the agent using the internal engine APIs
- API integration: Exposing logs using a similar API than the regular `docker logs`
- CLI integration: `docker service logs`

I want to validate the API before documenting and adding integration tests, so this PR does not include them and is not in a mergeable state.

/cc @aaronlehmann @stevvooe

### CODE REVIEW

File: client/service_logs.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	if options.Details {
+		query.Set("details", "1")
+	}
+
+	if options.Follow {
+		query.Set("follow", "1")
+	}
+	query.Set("tail", options.Tail)
+
+	resp, err := cli.get(ctx, "/services/"+serviceID+"/logs", query, nil)
```
Comment (aaronlehmann) [inline client/service_logs.go]: Wondering if we should validate client-side that user input only contains valid characters. Passing in a service ID with a `?` in it will cause some pretty weird behavior. But this should probably be done in a different PR.

File: api/server/router/swarm/cluster_routes.go [hunk 2]
```diff
@@ -173,6 +174,51 @@ func (sr *swarmRouter) removeService(ctx context.Context, w http.ResponseWriter,
...
+	// any error after the stream starts (i.e. container not found, wrong parameters)
+	// with the appropriate status code.
+	stdout, stderr := httputils.BoolValue(r, "stdout"), httputils.BoolValue(r, "stderr")
+	if !(stdout || stderr) {
+		return fmt.Errorf("Bad parameters: you must choose at least one stream")
+	}
+
+	serviceName := vars["id"]
+	logsConfig := &backend.ContainerLogsConfig{
+		ContainerLogsOptions: basictypes.ContainerLogsOptions{
```
Comment (stevvooe) [inline api/server/router/swarm/cluster_routes.go]: Do we rename this to `LogsOptions`?

File: daemon/cluster/cluster.go [hunk 3]
```diff
@@ -1018,6 +1026,79 @@ func (c *Cluster) RemoveService(input string) error {
...
+					return err
+				}
+				data = append([]byte(ts.Format(logger.TimeFormat)+" "), data...)
+			}
+
+			data = append([]byte(fmt.Sprintf("node_id=%s,service_id=%s,task_id=%s ",
+				msg.Context.NodeID,
+				msg.Context.ServiceID,
+				msg.Context.TaskID,
+			)), data...)
```
Comment (stevvooe) [inline daemon/cluster/cluster.go]: Actually, it might be more efficient to just flush to the stream, rather than assembling the entire data buffer.

Comment (aluzzardi) [inline daemon/cluster/cluster.go]: Followed the same pattern as `docker logs` (reverse append) - I can change that

**Comment by aluzzardi** [comment]
@aaronlehmann @stevvooe Actually, the biggest thing I need to solve is the format.

Here's docker logs:
```
[optional timestamp] [optional details] [line]
```

Details (`docker logs --details`) are key values (`k1=v1,k2=v2`) which I believe are set by the log driver?

In `service logs`, we have a context (`node_id`, `task_id`, etc).

- Should we support "details" in `service logs`?
- Should we inject the context into the details or keep it separate?

Having the context in details makes the service logs protocol 100% compliant with `docker logs`, although it gets a little messier for parsing: depending whether `--timestamp` and or `--details` was passed or not, we have to split the log stream differently. Details contains the context (which we need to resolve to actual service names) and the actual details (which we need to passthrough).

Thoughts?

/cc @tiborvass @icecrime @vieux

**Comment by vieux** [comment]
@aluzzardi I would start by the timestamp, so I vote for `3)` or `4)` details is nice to have but can come later.

**Comment by stevvooe** [comment]
> 3) [optional timestamp] [optional details + context] [line]

I vote for 3.

**Comment by aluzzardi** [comment]
Which one though:
```
[optional(details) + context]
OR
[optional(details + context)]
```

@dnephin @aanand Any input? The output of `service logs` basically looks a lot like `compose logs`.

**Comment by stevvooe** [comment]
> [optional(details) + context]

I'm in favor of this, since you _need_ the context to disambiguate the streams.

**Comment by vieux** [comment]
@aluzzardi #28088 was merged

File: cli/command/service/logs.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+type logWriter struct {
+	ctx context.Context
+	r   *idresolver.IDResolver
+	w   io.Writer
+}
+
+func (lw *logWriter) Write(buf []byte) (int, error) {
+	parts := bytes.SplitN(buf, []byte(" "), 2)
```
Comment (dongluochen) [inline cli/command/service/logs.go]: Does this mean context cannot have any space?

File: cli/command/service/logs.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/docker/docker/cli/command/idresolver"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/spf13/cobra"
+)
+
+type logsOptions struct {
+	noResolve  bool
+	follow     bool
+	since      string
+	timestamps bool
```
Comment (dongluochen) [inline cli/command/service/logs.go]: Does logs from different nodes ordered by timestamps by default? As an option?

**Comment by aluzzardi** [comment]
>> [optional(details) + context]

> I'm in favor of this, since you need the context to disambiguate the streams.

Keep in mind that this means:
1) Logs format is incompatible with service logs (because there's an extra column even if you don't provide `--details`)
2) CLI parsing would require to: Extract the "thing" in the middle, parse out the relevant part (node_id and so on) to resolve and display it, and if `--details` was provided, take what's remaining of that "thing" (minus node_id and so on) and display it

Are we cool with that, @stevvooe @aaronlehmann @vieux @icecrime ?

**Comment by dnephin** [comment]
I think the logs are going to be pretty confusing without the task id. I would opt for always including a task id by default (maybe a flag to disable it).

**Comment by aluzzardi** [comment]
PTAL @stevvooe @vieux @stevvooe 

There are a few things far from perfect (e.g. the reference docs are mostly copy paste from container logs) - could you please distinguish your comments between fix now and fix later? I'll make a follow-up with all less important fixes

**Comment by aluzzardi** [comment]
Stop gaps: non-follow is not supported, details is not supported

File: api/server/router/swarm/backend.go [hunk 6]
```diff
@@ -17,6 +19,7 @@ type Backend interface {
 	CreateService(types.ServiceSpec, string) (string, error)
 	UpdateService(string, uint64, types.ServiceSpec, string, string) error
 	RemoveService(string) error
+	ServiceLogs(context.Context, string, *backend.ContainerLogsConfig, chan struct{}) error
```
Comment (aluzzardi) [inline api/server/router/swarm/backend.go]: Staying consistent with `ContainerLogs`

File: api/server/router/swarm/cluster_routes.go (hunk 2 above)
Comment (aluzzardi) [inline api/server/router/swarm/cluster_routes.go]: In a follow up

File: api/server/router/swarm/backend.go (hunk 6 above)
Comment (stevvooe) [inline api/server/router/swarm/backend.go]: > It'd be better to follow Verb + Noun naming convention. 

This is not a convention in Go.

File: api/server/router/swarm/cluster_routes.go (hunk 2 above)
Comment (stevvooe) [inline api/server/router/swarm/cluster_routes.go]: Can we add a TODO somewhere?

File: cli/command/service/logs.go (hunk 5 above)
Comment (stevvooe) [inline cli/command/service/logs.go]: No. The resource requirements for doing so are too intensive. In practice, they tend to come in with the correct ordering. To guarantee it, you have to buffer all logs, sort them and merge on client-side. Empty signals also need to be sent to preserve ordering.

**Comment by vieux** [comment]
@aluzzardi I get

```
$ docker service logs pbf1duhxowx0
Error response from daemon: Bad parameters: Only follow mode is currently supported
```

File: cli/command/service/logs.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+	components := strings.Split(input, ",")
+	for _, component := range components {
+		parts := strings.SplitN(component, "=", 2)
+		if len(parts) != 2 {
+			return "", "", fmt.Errorf("invalid context: %s", input)
+		}
+		context[parts[0]] = parts[1]
+	}
+
+	taskID, ok := context["task_id"]
```
Comment (stevvooe) [inline cli/command/service/logs.go]: We should use the convention of `task.id` we've used in labels. It might also be a good idea to namespace these `com.docker.swarm`.

**Comment by aluzzardi** [comment]
@vieux Yeah, unfortunately, only follow mode (`-f`) is supported

**Comment by vieux** [comment]
my bad @aluzzardi I read `non-follow is supported, details is not supported`

**Comment by vieux** [comment]
@aluzzardi works great with `-f` though :)

**Comment by vieux** [comment]
neat 👍 

```
$ docker service create --replicas 3 redis
pbf1duhxowx0z66ma4y2oq1yb

$ docker service logs -f pbf1duhxowx0
jovial_bhabha.2.nsj65viveh8t@dev    | 1:C 10 Nov 02:13:28.502 # Warning: no config file specified, using the default config. In order to specify a config file use redis-server /path/to/redis.conf
jovial_bhabha.3.0emnvq7gr6iu@dev    | 1:C 10 Nov 02:13:28.435 # Warning: no config file specified, using the default config. In order to specify a config file use redis-server /path/to/redis.conf
jovial_bhabha.1.ex9jdel9krtn@dev    | 1:C 10 Nov 02:13:28.594 # Warning: no config file specified, using the default config. In order to specify a config file use redis-server /path/to/redis.conf
jovial_bhabha.1.ex9jdel9krtn@dev    |                 _._
jovial_bhabha.2.nsj65viveh8t@dev    |                 _._
jovial_bhabha.3.0emnvq7gr6iu@dev    |                 _._
jovial_bhabha.3.0emnvq7gr6iu@dev    |            _.-``__ ''-._
jovial_bhabha.1.ex9jdel9krtn@dev    |            _.-``__ ''-._
jovial_bhabha.2.nsj65viveh8t@dev    |            _.-``__ ''-._
jovial_bhabha.3.0emnvq7gr6iu@dev    |       _.-``    `.  `_.  ''-._           Redis 3.2.5 (00000000/0) 64 bit
jovial_bhabha.2.nsj65viveh8t@dev    |       _.-``    `.  `_.  ''-._           Redis 3.2.5 (00000000/0) 64 bit
jovial_bhabha.1.ex9jdel9krtn@dev    |       _.-``    `.  `_.  ''-._           Redis 3.2.5 (00000000/0) 64 bit
jovial_bhabha.1.ex9jdel9krtn@dev    |   .-`` .-```.  ```\/    _.,_ ''-._
jovial_bhabha.3.0emnvq7gr6iu@dev    |   .-`` .-```.  ```\/    _.,_ ''-._
jovial_bhabha.2.nsj65viveh8t@dev    |   .-`` .-```.  ```\/    _.,_ ''-._
jovial_bhabha.2.nsj65viveh8t@dev    |  (    '      ,       .-`  | `,    )     Running in standalone mode
jovial_bhabha.1.ex9jdel9krtn@dev    |  (    '      ,       .-`  | `,    )     Running in standalone mode
jovial_bhabha.3.0emnvq7gr6iu@dev    |  (    '      ,       .-`  | `,    )     Running in standalone mode
jovial_bhabha.3.0emnvq7gr6iu@dev    |  |`-._`-...-` __...-.``-._|'` _.-'|     Port: 6379
jovial_bhabha.2.nsj65viveh8t@dev    |  |`-._`-...-` __...-.``-._|'` _.-'|     Port: 6379
jovial_bhabha.1.ex9jdel9krtn@dev    |  |`-._`-...-` __...-.``-._|'` _.-'|     Port: 6379
jovial_bhabha.2.nsj65viveh8t@dev    |  |    `-._   `._    /     _.-'    |     PID: 1
jovial_bhabha.3.0emnvq7gr6iu@dev    |  |    `-._   `._    /     _.-'    |     PID: 1
jovial_bhabha.3.0emnvq7gr6iu@dev    |   `-._    `-._  `-./  _.-'    _.-'
jovial_bhabha.1.ex9jdel9krtn@dev    |  |    `-._   `._    /     _.-'    |     PID: 1
jovial_bhabha.2.nsj65viveh8t@dev    |   `-._    `-._  `-./  _.-'    _.-'
jovial_bhabha.2.nsj65viveh8t@dev    |  |`-._`-._    `-.__.-'    _.-'_.-'|
jovial_bhabha.3.0emnvq7gr6iu@dev    |  |`-._`-._    `-.__.-'    _.-'_.-'|
jovial_bhabha.1.ex9jdel9krtn@dev    |   `-._    `-._  `-./  _.-'    _.-'
jovial_bhabha.1.ex9jdel9krtn@dev    |  |`-._`-._    `-.__.-'    _.-'_.-'|
jovial_bhabha.2.nsj65viveh8t@dev    |  |    `-._`-._        _.-'_.-'    |           http://redis.io
jovial_bhabha.3.0emnvq7gr6iu@dev    |  |    `-._`-._        _.-'_.-'    |           http://redis.io
jovial_bhabha.3.0emnvq7gr6iu@dev    |   `-._    `-._`-.__.-'_.-'    _.-'
jovial_bhabha.1.ex9jdel9krtn@dev    |  |    `-._`-._        _.-'_.-'    |           http://redis.io
jovial_bhabha.2.nsj65viveh8t@dev    |   `-._    `-._`-.__.-'_.-'    _.-'
jovial_bhabha.2.nsj65viveh8t@dev    |  |`-._`-._    `-.__.-'    _.-'_.-'|
jovial_bhabha.3.0emnvq7gr6iu@dev    |  |`-._`-._    `-.__.-'    _.-'_.-'|
jovial_bhabha.1.ex9jdel9krtn@dev    |   `-._    `-._`-.__.-'_.-'    _.-'
jovial_bhabha.2.nsj65viveh8t@dev    |  |    `-._`-._        _.-'_.-'    |
jovial_bhabha.2.nsj65viveh8t@dev    |   `-._    `-._`-.__.-'_.-'    _.-'
jovial_bhabha.3.0emnvq7gr6iu@dev    |  |    `-._`-._        _.-'_.-'    |
jovial_bhabha.1.ex9jdel9krtn@dev    |  |`-._`-._    `-.__.-'    _.-'_.-'|
jovial_bhabha.1.ex9jdel9krtn@dev    |  |    `-._`-._        _.-'_.-'    |
jovial_bhabha.2.nsj65viveh8t@dev    |       `-._    `-.__.-'    _.-'
jovial_bhabha.3.0emnvq7gr6iu@dev    |   `-._    `-._`-.__.-'_.-'    _.-'
jovial_bhabha.3.0emnvq7gr6iu@dev    |       `-._    `-.__.-'    _.-'
jovial_bhabha.1.ex9jdel9krtn@dev    |   `-._    `-._`-.__.-'_.-'    _.-'
jovial_bhabha.2.nsj65viveh8t@dev    |           `-._        _.-'
jovial_bhabha.2.nsj65viveh8t@dev    |               `-.__.-'
jovial_bhabha.3.0emnvq7gr6iu@dev    |           `-._        _.-'
jovial_bhabha.1.ex9jdel9krtn@dev    |       `-._    `-.__.-'    _.-'
jovial_bhabha.1.ex9jdel9krtn@dev    |           `-._        _.-'
jovial_bhabha.2.nsj65viveh8t@dev    |
jovial_bhabha.3.0emnvq7gr6iu@dev    |               `-.__.-'
jovial_bhabha.3.0emnvq7gr6iu@dev    |
jovial_bhabha.1.ex9jdel9krtn@dev    |               `-.__.-'
jovial_bhabha.2.nsj65viveh8t@dev    | 1:M 10 Nov 02:13:28.508 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
jovial_bhabha.3.0emnvq7gr6iu@dev    | 1:M 10 Nov 02:13:28.438 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
jovial_bhabha.1.ex9jdel9krtn@dev    |
jovial_bhabha.2.nsj65viveh8t@dev    | 1:M 10 Nov 02:13:28.508 # Server started, Redis version 3.2.5
jovial_bhabha.2.nsj65viveh8t@dev    | 1:M 10 Nov 02:13:28.508 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
jovial_bhabha.3.0emnvq7gr6iu@dev    | 1:M 10 Nov 02:13:28.439 # Server started, Redis version 3.2.5
jovial_bhabha.1.ex9jdel9krtn@dev    | 1:M 10 Nov 02:13:28.597 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
jovial_bhabha.1.ex9jdel9krtn@dev    | 1:M 10 Nov 02:13:28.597 # Server started, Redis version 3.2.5
jovial_bhabha.2.nsj65viveh8t@dev    | 1:M 10 Nov 02:13:28.508 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
jovial_bhabha.3.0emnvq7gr6iu@dev    | 1:M 10 Nov 02:13:28.439 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
jovial_bhabha.2.nsj65viveh8t@dev    | 1:M 10 Nov 02:13:28.508 * The server is now ready to accept connections on port 6379
jovial_bhabha.1.ex9jdel9krtn@dev    | 1:M 10 Nov 02:13:28.597 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
jovial_bhabha.1.ex9jdel9krtn@dev    | 1:M 10 Nov 02:13:28.597 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
jovial_bhabha.3.0emnvq7gr6iu@dev    | 1:M 10 Nov 02:13:28.439 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
jovial_bhabha.3.0emnvq7gr6iu@dev    | 1:M 10 Nov 02:13:28.439 * The server is now ready to accept connections on port 6379
jovial_bhabha.1.ex9jdel9krtn@dev    | 1:M 10 Nov 02:13:28.597 * The server is now ready to accept connections on port 6379

```

**Comment by stevvooe** [comment]
LGTM

I think we can tweak the output quite a bit but let's do that in a follow up.

File: daemon/cluster/cluster.go [hunk 8]
```diff
@@ -1121,6 +1130,82 @@ func (c *Cluster) RemoveService(input string) error {
 	return nil
 }
 
+// ServiceLogs collects service logs and writes them back to `config.OutStream`
+func (c *Cluster) ServiceLogs(ctx context.Context, input string, config *backend.ContainerLogsConfig, started chan struct{}) error {
+	if !c.isActiveManager() {
```
Comment (tonistiigi) [inline daemon/cluster/cluster.go]: Need to get a readlock for this. You can release it before the long running operations.

Comment (aluzzardi) [inline daemon/cluster/cluster.go]: Done. PTAL

**Comment by vieux** [comment]
It would be nice to get it for 1.13, but since it's still missing a few feature and after some discussion among the engine maintainers, we think it's best if you can put this in experimental only @al

File: daemon/cluster/cluster.go (hunk 8 above)
Comment (tonistiigi) [inline daemon/cluster/cluster.go]: LGTM

**Comment by aluzzardi** [comment]
Moved to experimental

@vieux PTAL

**Comment by vieux** [comment]
LGTM

**Comment by bvis** [comment]
Hi,

I tried this feature in my staging swarm cluster and it didn't worked as expected. I reproduced the error here: http://play-with-docker.com/p/f0020045-d8e8-4bf0-a86c-55f3a25cd4dd

I left some messages related with this in the swam slack room.

If you want me to create an issue just tell it to me.

**Comment by aaronlehmann** [comment]
@bvis: I wasn't able to load your link. I get the error message "session not found".

Opening an issue with some details sounds like a good next step.

**Comment by bvis** [comment]
@aaronlehmann Here you are: https://github.com/docker/docker/issues/28915
